misc/scripts/models-as-data/rust_bulk_generate_mad.py

```diff
@@ -0,0 +1,335 @@
+        "git_tag": "0.2.172",
+    },
+    {
+        "name": "log",
+        "git_repo": "https://github.com/rust-lang/log",
+        "git_tag": "0.4.27",
+    },
+    {
+        "name": "memchr",
+        "git_repo": "https://github.com/BurntSushi/memchr",
+        "git_tag": "2.7.4",
+    },
+    {
+        "name": "once_cell",
+        "git_repo": "https://github.com/matklad/once_cell",
+        "git_tag": "v1.21.3",
+    },
+    {
+        "name": "rand",
+        "git_repo": "https://github.com/rust-random/rand",
+        "git_tag": "0.9.1",
+    },
+    {
+        "name": "smallvec",
+        "git_repo": "https://github.com/servo/rust-smallvec",
+        "git_tag": "v1.15.0",
+    },
+    {
+        "name": "serde",
+        "git_repo": "https://github.com/serde-rs/serde",
+        "git_tag": "v1.0.219",
+    },
+    {
+        "name": "tokio",
+        "git_repo": "https://github.com/tokio-rs/tokio",
+        "git_tag": "tokio-1.45.0",
+    },
+    {
+        "name": "reqwest",
+        "git_repo": "https://github.com/seanmonstar/reqwest",
+        "git_tag": "v0.12.15",
+    },
+    {
+        "name": "rocket",
+        "git_repo": "https://github.com/SergioBenitez/Rocket",
+        "git_tag": "v0.5.1",
+    },
+    {
+        "name": "actix-web",
+        "git_repo": "https://github.com/actix/actix-web",
+        "git_tag": "web-v4.11.0",
+    },
+    {
+        "name": "hyper",
+        "git_repo": "https://github.com/hyperium/hyper",
+        "git_tag": "v1.6.0",
+    },
+    {
+        "name": "clap",
+        "git_repo": "https://github.com/clap-rs/clap",
+        "git_tag": "v4.5.38",
+    },
+]
+
+
+def clone_project(project: Project) -> str:
+    """
+    Shallow clone a project into the build directory.
+
+    Args:
+        project: A dictionary containing project information with 'name', 'git_repo', and optional 'git_tag' keys.
+
+    Returns:
+        The path to the cloned project directory.
+    """
+    name = project["name"]
+    repo_url = project["git_repo"]
+    git_tag = project.get("git_tag")
+
+    # Determine target directory
+    target_dir = os.path.join(build_dir, name)
+
+    # Clone only if directory doesn't already exist
+    if not os.path.exists(target_dir):
+        if git_tag:
+            print(f"Cloning {name} from {repo_url} at tag {git_tag}")
+        else:
+            print(f"Cloning {name} from {repo_url}")
+
+        subprocess.check_call(
+            [
+                "git",
+                "clone",
+                "--quiet",
+                "--depth",
+                "1",  # Shallow clone
+                *(
+                    ["--branch", git_tag] if git_tag else []
+                ),  # Add branch if tag is provided
+                repo_url,
+                target_dir,
+            ]
+        )
+        print(f"Completed cloning {name}")
+    else:
+        print(f"Skipping cloning {name} as it already exists at {target_dir}")
```

Comment:
Ideally at some point we should handle the case that the specified `git_tag` has been updated (by fetching the new version).